### DISCUSSION TITLE
Python Behavior Tree  for ROS2 ?

### DISCUSSION THREAD

**Discussion by framlin** [iRobotEducation/create3_docs#285] [Questions] [answered]
Hello, I have a general question about ROS2.
I have bought the create3, to learn ROS2. Before that, I built a view little mobile robots and found out, that it is useful to have something like a behavior tree, to control the robots activities. For my self made robots I implemented them for myself, but now I want to use ROS2-tools.
Is there a Python BT-lib for galactic? My first tries to search for them were not successful, but I cannot imagine, that galactic does not provide a working BT-lib. 
Does anybody know, were to fine one and if there are tutorials, how to use it?

**Comment by alsora** [comment, marked as answer]
IMO it's a fallacy to look for a "behavior tree library for ROS 2".

ROS is a communication framework, while a behavior tree is a model to plan the execution of tasks.
The main purpose of ROS is to abstract the connection between different components, but it does not care what these components do internally.
You don't need a "behavior tree library for ROS 2", but rather you need to take any generic behavior tree library and "wrap it" inside ROS 2 APIs.
Writing a behavior tree library that is heavily dependent on ROS and meant to be only used through it would be a bad design choice and that's why you will likely not find it.

Having said that, considering the popularity of ROS 2 and the recent increased adoption of behavior trees in robotics, there are existing implementations of such "wrappers".
But I really want to stress that, depending on your purpose and application, it may be sufficient to use an extremely simple and thin wrapper as its purpose will mostly be to reduce boilerplate when interacting with the underlying behavior tree library.
Building such wrapper by yourself would be a great learning opportunity, both to understand how to integrate libraries in ROS as well as getting to know better behavior trees.

Here are some examples that I know of:
 - [BehaviorTree.cpp](https://github.com/BehaviorTree/BehaviorTree.CPP) this is a C++ library, not Python, but it's probably the most complete and widely used behavior tree library for robotics. It's used by the ROS 2 navigation stack, where you can find an example of this "wrapper" https://github.com/ros-planning/navigation2/tree/main/nav2_behavior_tree
 - [py_tees](https://github.com/splintered-reality/py_trees) this is a Python BT library. It has a [ROS wrapper](https://github.com/splintered-reality/py_trees_ros) that, although it officially supports only up to ROS 2 Foxy, it can also be used in Galactic if downloaded and built from sources, as mentioned by the author himself https://github.com/splintered-reality/py_trees_ros/issues/174#issuecomment-1142155133

**Reply by framlin** [reply]
Thank you very much for pointing out my misunderstanding!
It sound's reasonable, that it might be a good training, to write myself a simple wrapper.

That leads to a spontaneous new question coming up in my mind.
Can I interrupt an ongoing action? For example a drive_distance - action, if the ir-sensor suddenly detects an obstacle?

**Reply by alsora** [reply]
Yes, one of the key characteristics of ROS 2 actions is that they can be canceled by the client while they are executing.

A client sends a goal request and it receives as ouput a goal handle object.
This is used to monitor the action execution and to eventually retrieve the result when available.
It can also be used to cancel the goal, using this method https://github.com/ros2/rclpy/blob/rolling/rclpy/rclpy/action/client.py#L72

Unfortunately a quick Google search didn't show a good tutorial on how to do that in python, but it's relatively straight forward (I.e. Just call that method)

**Reply by framlin** [reply]
Thank you very much for your help!! Great!

**Reply by alsora** [reply]
I stumbled upon this repository with some examples of using behavior trees in both pytjon and c++.

https://github.com/sea-bass/turtlebot3_behavior_demos
